ledger/onlineaccountscache.go

```diff
@@ -0,0 +1,136 @@
+	accounts map[basics.Address]*persistedOnlineAccountDataList
+}
+
+// init initializes the onlineAccountsCache for use.
+// thread locking semantics : write lock
+func (o *onlineAccountsCache) init(accts []persistedOnlineAccountData) {
+	o.accounts = make(map[basics.Address]*persistedOnlineAccountDataList)
+	for _, acct := range accts {
+		// if cache full, stop writing
+		if !o.writeFront(acct) {
+			break
+		}
+	}
+}
+
+func (o *onlineAccountsCache) full() bool {
+	return len(o.accounts) >= onlineAccountsCacheMaxSize
+}
+
+// read the persistedAccountData object that the cache has for the given address.
+// thread locking semantics : read lock
+func (o *onlineAccountsCache) read(addr basics.Address, rnd basics.Round) (persistedOnlineAccountData, bool) {
+	if list := o.accounts[addr]; list != nil {
+		node := list.back()
+		if node.Value.updRound > rnd {
+			return persistedOnlineAccountData{}, false
+		}
+		for node.prev != &list.root {
+			node = node.prev
+			// only need one entry that is targetRound or older
+			if node.Value.updRound > rnd {
+				return *node.next.Value, true
+			}
+		}
+		return *node.Value, true
+	}
+	return persistedOnlineAccountData{}, false
+}
+
+// write a single persistedAccountData to the cache
+// thread locking semantics : write lock
+func (o *onlineAccountsCache) writeFront(acctData persistedOnlineAccountData) bool {
+	if _, ok := o.accounts[acctData.addr]; !ok {
+		if o.full() {
+			return false
+		}
+		o.accounts[acctData.addr] = newPersistedOnlineAccountList()
+	}
+	list := o.accounts[acctData.addr]
+	if list.root.next != &list.root && acctData.updRound <= list.root.next.Value.updRound {
+		return false
+	}
+	o.accounts[acctData.addr].pushFront(&acctData)
+	return true
+}
+
+// write a single persistedAccountData to the cache
+// thread locking semantics : write lock
+func (o *onlineAccountsCache) writeBack(acctData persistedOnlineAccountData) bool {
+	if _, ok := o.accounts[acctData.addr]; !ok {
+		if o.full() {
+			return false
+		}
+		o.accounts[acctData.addr] = newPersistedOnlineAccountList()
+	}
+	list := o.accounts[acctData.addr]
+	if list.root.prev != &list.root && acctData.updRound >= list.root.prev.Value.updRound {
+		return false
+	}
+	o.accounts[acctData.addr].pushBack(&acctData)
+	return true
+}
+
+// prune trims the onlineaccountscache by only keeping entries that would give account state
+// of rounds targetRound and later
+// thread locking semantics : write lock
+func (o *onlineAccountsCache) prune(targetRound basics.Round) {
+	for addr, list := range o.accounts {
+		node := list.back()
+		for node.prev != &list.root {
+			node = node.prev
+			// keep only one entry that is targetRound or older
+			// discard all older additional entries older than targetRound
+			if node.Value.updRound <= targetRound {
+				list.remove(node.next)
+			} else {
+				break
+			}
+		}
+		// only one item left in cache
+		if node.prev == &list.root && node.next == &list.root {
```

Comment:
`node.next == &list.root` is always true I think

Reply:
its not true when the break statement was reached.